macrosAutoCats/src/main/scala/com/softwaremill/macwire/autocats/internals/CatsProviders.scala

```diff
@@ -0,0 +1,147 @@
+    def maybeUnderlyingType(tpe: Type): Option[Type] = if (isResource(tpe)) Some(underlyingType(tpe)) else None
+
+    def isResource(tpe: Type): Boolean =
+      tpe.typeSymbol.fullName.startsWith("cats.effect.kernel.Resource") && tpe.typeArgs.size == 2
+
+    def fromTree(tree: Tree, tpe: Type): Option[Resource] =
+      if (isResource(tpe)) Some(new Resource(tree))
+      else None
+
+  }
+
+  class FactoryMethod(fun: c.Tree, val resultType: c.Type, val dependencies: List[List[Option[Provider]]])
+      extends Provider {
+    import c.universe._
+
+    private lazy val appliedTree: Tree =
+      log.withBlock(s"Applied tree for [$fun] from deps: [${dependencies.map(_.mkString(", ")).mkString(", ")}]") {
+        dependencies.map(_.map(_.get.ident)).foldLeft(fun)((acc: Tree, args: List[Tree]) => Apply(acc, args))
+      }
+    lazy val result: Provider = log.withResult {
+      val t = fun.symbol.asMethod.returnType
+
+      //TODO support for FactoryMethods
+      if (Resource.isResource(t)) new Resource(appliedTree) {
+        override lazy val resultType: Type = Resource.underlyingType(t)
+      }
+      else if (Effect.isEffect(t)) new Effect(appliedTree) {
+        override lazy val resultType: Type = Effect.underlyingType(t)
+      }
+      else new Instance(appliedTree)
+    }(result => s"Factory method result [$result]")
+
+    lazy val ident: Tree = log.withResult(result.ident)(i => s"Ident for [$fun] is [$i] object: [$this]")
+
+    lazy val value: Tree = result.value
+
+  }
+
+  object FactoryMethod {
+    import c.universe._
+
+    def underlyingType(tree: Tree): Type = {
+      val resultType = tree.symbol.asMethod.returnType
+      (Resource.maybeUnderlyingType(resultType) orElse Effect.maybeUnderlyingType(resultType)).getOrElse(resultType)
+    }
+
+    def underlyingResultType(tree: Tree): Type = {
+      val (fun, _) =
+        fromTree(tree).getOrElse(c.abort(c.enclosingPosition, s"The given tree is not a factory method: [$tree]"))
+      underlyingType(fun)
+    }
+
+    def unapply(tree: Tree): Option[(Tree, List[ValDef])] = deconstruct(identity)(tree)
+    def fromTree(tree: Tree): Option[(Tree, List[ValDef])] = unapply(tree)
+    def isFactoryMethod(tree: Tree): Boolean = fromTree(tree).isDefined
+
+    def deconstruct[T](componentsTransformer: ((Tree, List[ValDef])) => T)(tree: Tree): Option[T] = tree match {
+      // Function with two parameter lists (implicit parameters) (<2.13)
+      case Block(Nil, Function(p, Apply(Apply(f, _), _))) => Some(componentsTransformer((f, p)))
+      case Block(Nil, Function(p, Apply(f, _)))           => Some(componentsTransformer((f, p)))
+      // Function with two parameter lists (implicit parameters) (>=2.13)
+      case Function(p, Apply(Apply(f, _), _)) => Some(componentsTransformer((f, p)))
+      case Function(p, Apply(f, _))           => Some(componentsTransformer((f, p)))
+      // Other types not supported
+      case _ => None
+    }
+
+  }
+
+  case class Creator(
```

Comment:
what's the difference between a creator and a factory? isn't a constructor/apply method just a special case of factory?

Reply:
Ah there is no real difference in terms of the dependencies graph. I was focused on the different sources of these providers and missed that they are, actually the same... Thanks for the comment!